training/mochi-1/text_to_video_lora.py

```diff
@@ -0,0 +1,538 @@
+        if isinstance(module, torch.nn.Linear):
+            assert any(
+                n in name for n in ["time_embed", "proj_out", "blocks", "norm_out"]
+            ), f"Unexpected linear layer: {name}"
+            module.to(dtype=dtype)
+        elif isinstance(module, torch.nn.Conv2d):
+            module.to(dtype=dtype)
+    return model
+
+
+class CollateFunction:
+    def __init__(self, caption_dropout: float = None) -> None:
+        self.caption_dropout = caption_dropout
+
+    def __call__(self, samples: List[Tuple[dict, torch.Tensor]]) -> Dict[str, torch.Tensor]:
+        ldists = torch.cat([data[0]["ldist"] for data in samples], dim=0)
+        z = DiagonalGaussianDistribution(ldists).sample()
+        assert torch.isfinite(z).all()
+
+        # Sample noise which we will add to the samples.
+        eps = torch.randn_like(z)
+        sigma = torch.rand(z.shape[:1], device="cpu", dtype=torch.float32)
+
+        prompt_embeds = torch.cat([data[1]["prompt_embeds"] for data in samples], dim=0)
+        prompt_attention_mask = torch.cat([data[1]["prompt_attention_mask"] for data in samples], dim=0)
+        if self.caption_dropout and random.random() < self.caption_dropout:
+            prompt_embeds.zero_()
+            prompt_attention_mask = prompt_attention_mask.long()
+            prompt_attention_mask.zero_()
+            prompt_attention_mask = prompt_attention_mask.bool()
+
+        return dict(
+            z=z, eps=eps, sigma=sigma, prompt_embeds=prompt_embeds, prompt_attention_mask=prompt_attention_mask
+        )
+
+
+def main(args):
+    if not torch.cuda.is_available():
+        raise ValueError("Not supported without CUDA.")
+    
+    if args.report_to == "wandb" and args.hub_token is not None:
+        raise ValueError(
+            "You cannot use both --report_to=wandb and --hub_token due to a security risk of exposing your token."
+            " Please use `huggingface-cli login` to authenticate with the Hub."
+        )
+
+    # Handle the repository creation
+    if args.output_dir is not None:
+        os.makedirs(args.output_dir, exist_ok=True)
+
+    if args.push_to_hub:
+        repo_id = create_repo(
+            repo_id=args.hub_model_id or Path(args.output_dir).name,
+            exist_ok=True,
+        ).repo_id
+
+    # Prepare models and scheduler
+    transformer = MochiTransformer3DModel.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="transformer",
+        revision=args.revision,
+        variant=args.variant,
+    )
+    scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="scheduler"
+    )
+
+    transformer.requires_grad_(False)
+    transformer.to("cuda")
+    if args.gradient_checkpointing:
+        transformer.enable_gradient_checkpointing()
+    if args.cast_dit:
+        transformer = cast_dit(transformer, torch.bfloat16)
+    if args.compile_dit:
+        transformer.compile()
```

Comment:
Interesting... I did not know we had a method that allows compilation this way. This is torch.compile-ation right?

Reply:
Yes sir.